framework/include/postprocessors/Postprocessor.h

```diff
@@ -10,39 +10,119 @@
 #pragma once
 
 #include "OutputInterface.h"
-#include "FEProblemBase.h"
 #include "NonADFunctorInterface.h"
 #include "libmesh/parallel.h"
 
+class MooseObject;
+
 /**
  * Base class for all Postprocessors.  Defines a name and sets up the
  * virtual getValue() interface which must be overridden by derived
  * classes.
  */
-class Postprocessor : public OutputInterface, public NonADFunctorInterface
+class Postprocessor : public OutputInterface,
+                      public NonADFunctorInterface,
+                      public Moose::FunctorBase<Real>
 {
 public:
   static InputParameters validParams();
 
   Postprocessor(const MooseObject * moose_object);
 
   /**
-   * This will get called to actually grab the final value the postprocessor has calculated.
+   * This will get called to actually grab the final value the postprocessor has calculated
+   *
+   * Note that this should only be called by internal methods, namely the problem that
+   * actually sets the value globally for other things to use. If you want the value
+   * outside of one of these external methods, you should use getCurrentValue().
    *
    * This method will be removed in favor of the const version.
    */
   virtual PostprocessorValue getValue();
 
   /**
    * This will get called to actually grab the final value the postprocessor has calculated.
+   *
+   * Note that this should only be called by internal methods, namely the problem that
+   * actually sets the value globally for other things to use. If you want the value
+   * outside of one of these external methods, you should use getCurrentValue().
    */
   virtual PostprocessorValue getValue() const;
 
+  /**
+   * @return The "current" value of this Postprocessor.
+   *
+   * Your sanity would tell you... why not just call getValue()? Well - the intention
+   * of getValue() is to be called by the problem when the UserObjects are executed,
+   * and not by other things. This enables the control of _when_ this Postprocessor
+   * is updated, which could be very important. If the implementation of getValue() is
+   * such that it actually computes a new value (instead of one that is called in
+   * finalize()), you could potentially call getValue() and not get the value as it
+   * was at the last time this PP was executed.
+   *
+   * What this does instead is gives you the value that was last set as this PP was
+   * executed by the problem. That is, the value that every object that uses the
+   * PostprocessorInterface will get you.
+   */
+  const PostprocessorValue & getCurrentValue() const { return _current_value; }
+
   /**
    * Returns the name of the Postprocessor.
    */
-  std::string PPName() const { return _pp_name; }
+  const std::string & PPName() const { return _pp_name; }
 
 protected:
-  const std::string _pp_name;
+  /// MOOSE object
+  const MooseObject & _pp_moose_object;
```

Comment:
umm this should probably be private? we dont want users working with this do we?

Reply:
There's no harm as it's typically the same object you're in, but I do like the idea of making it private to get rid of the clutter